File: go/21.merge-two-sorted-lists.go
/*
 * @lc app=leetcode id=21 lang=golang
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	hh := new(ListNode)
	h := hh
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			h.Next = l1
			l1 = l1.Next
		} else {
			h.Next = l2
			l2 = l2.Next
		}
		h = h.Next
	}
	if l1 == nil {
		h.Next = l2
	} else {
		h.Next = l1
	}
	return hh.Next
}

// @lc code=end

File: java/21.merge-two-sorted-lists.java
/*
 * @lc app=leetcode id=21 lang=java
 *
 * [21] Merge Two Sorted Lists
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        ListNode be4Head = new ListNode(0), ptr = be4Head;
        while (true) {
            if (l1 == null) {
                ptr.next = l2;
                break;
            }
            if (l2 == null) {
                ptr.next = l1;
                break;
            }
            if (l1.val < l2.val) {
                ptr.next = l1;
                ptr = l1;
                l1 = l1.next;
            } else {
                ptr.next = l2;
                ptr = l2;
                l2 = l2.next;
            }
        }
        return be4Head.next;
    }
}

